src/goto-harness/goto_harness_parse_options.cpp

```diff
@@ -7,46 +7,177 @@ Author: Diffblue Ltd.
+    << "--harness-function-name    the name of the harness function to "
+       "generate\n"
+    << "--harness-type             one of the harness types listed below\n"
+    << "\n\n"
+    << FUNCTION_HARNESS_GENERATOR_HELP << eom;
 }
 
 goto_harness_parse_optionst::goto_harness_parse_optionst(
   int argc,
   const char *argv[])
   : parse_options_baset{GOTO_HARNESS_OPTIONS, argc, argv, ui_message_handler},
+    messaget(ui_message_handler),
     ui_message_handler(cmdline, std::string("GOTO-HARNESS ") + CBMC_VERSION)
 {
 }
+
+goto_harness_parse_optionst::goto_harness_configt
+goto_harness_parse_optionst::handle_common_options()
+{
+  goto_harness_configt config{};
+
+  // This just checks the positional arguments to be 2.
+  // Options are not in .args
+  if(cmdline.args.size() != 2)
+  {
+    help();
+    throw invalid_command_line_argument_exceptiont{
+      "need to specify both input and output goto binary file names (may be "
+      "the same)",
+      "<in goto binary> <out goto binary>"};
+  }
+
+  config.in_file = cmdline.args[0];
+  config.out_file = cmdline.args[1];
+
+  if(!cmdline.isset(GOTO_HARNESS_GENERATOR_TYPE_OPT))
+  {
+    throw invalid_command_line_argument_exceptiont{
+      "required option not set", "--" GOTO_HARNESS_GENERATOR_TYPE_OPT};
+  }
+  config.harness_type = cmdline.get_value(GOTO_HARNESS_GENERATOR_TYPE_OPT);
+
+  // Read the name of the harness function to generate
+  if(!cmdline.isset(GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT))
+  {
+    throw invalid_command_line_argument_exceptiont{
+      "required option not set",
+      "--" GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT};
+  }
+  config.harness_function_name = irep_idt{
+    cmdline.get_value(GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT)};
+
+  return config;
+}
+
+void goto_harness_parse_optionst::invoke_harness(
+  goto_modelt &goto_model,
+  goto_harness_generatort &harness_generator,
+  const irep_idt &harness_function_name)
+{
+  auto const common_options =
+    std::set<std::string>{"version",
+                          GOTO_HARNESS_GENERATOR_TYPE_OPT,
+                          GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT};
+
+  for(auto const &option : cmdline.option_names())
+  {
+    if(common_options.find(option) == common_options.end())
+    {
+      harness_generator.handle_option(option, cmdline);
+    }
+  }
+
+  harness_generator.generate(goto_model, harness_function_name);
+  auto const &generated_harness =
+    goto_model.symbol_table.lookup_ref(harness_function_name);
+  goto_model.goto_functions.function_map[harness_function_name].type =
+    to_code_type(generated_harness.type);
+  auto &body =
+    goto_model.goto_functions.function_map[harness_function_name].body;
+  goto_convert(
+    static_cast<const codet &>(generated_harness.value),
+    goto_model.symbol_table,
+    body,
+    ui_message_handler,
+    ID_C);
+  body.add_instruction(goto_program_instruction_typet::END_FUNCTION);
```

Comment:
Why are the modification to `goto_functions` here and the modification to `symbol_table` in `generate` above.

Reply:
Good point, we thought that this is something that all generators would want to do in the end so we'd moved it here. After some further discussion it seems like this is not so generic a need as we thought, so we'll move this into the `function-call` generator instead